### PR TITLE
[proposal – catalyst.data] Paths dataset that derives labels from filenames

### PR DESCRIPTION
**Pull request by i-zaitsev**
Recently I decided to try `catalyst` in my projects, and I encountered the following case. In one of the datasets I am working on, the samples include labels as a part of file paths, like this.
```
\folder
    \with
        \samples
            \filename1_1.png
            \filename2_1.png
            \filename3_3.png
            \filename4_2.png
            ...
```
In `fastai` there are various datasets factories that can read features and targets from data frame columns, file paths, using lambda functions, etc. I decided to implement something similar for my case. Then, you can do as follows.
```python
import os
import re
import PIL.Image

class RegexLabelExtractor:
    def __init__(self, regex):
        self.regex = re.compile(regex)

    def __call__(self, filename):
        return int(self.regex.match(os.path.basename(filename)).group(1))

def list_files(folder):
    dirname = os.path.expanduser(folder)
    return [os.path.join(dirname, x) for x in os.listdir(dirname)]

def open_image(dict_):
    dict_['features'] = PIL.Image.open(dict_['features'])
    return dict_

dataset = FilesDataset(
    filenames=list_files('~/folder/with/samples'),
    open_fn=open_image,
    get_label_fn=RegexLabelExtractor('.*_(\d+)\\.png$'),
)    
```
In this PR, I propose to include something like this into `catalyst`. This implementation is pretty rough, so probably a better, or more generic solution could be implemented. (I am not sure if the proposed dataset implementation fits to the logic of the `catalyst` codebase). What do you think?

Thank you.

### CODE REVIEW

**Comment by TezRomacH** [comment]
Awesome!

File: catalyst/data/dataset.py [hunk 1]
```diff
@@ -112,4 +112,28 @@ def __len__(self) -> int:
...
+    Dataset that derives features and targets from samples filesystem paths.
+    """
+    def __init__(
+        self,
+        filenames: List[Dict],
+        open_fn: Callable,
+        get_label_fn: Callable,
+        **list_dataset_params
+    ):
+        list_data = [
```
Comment (TezRomacH) [inline catalyst/data/dataset.py:143]: Please, rewrite it like

```python
list_data = [
    dict(features=filename, targets=get_label_fn(filename))
    for filename in filenames
]
```

**Comment by i-zaitsev** [comment]
Updated the PR, now should look a bit better, I believe.

File: catalyst/data/dataset.py (hunk 1 above)
Comment (i-zaitsev) [reply catalyst/data/dataset.py:143]: Done!

**Comment by TezRomacH** [comment]
The build is still failing at the code-style stage
https://travis-ci.com/catalyst-team/catalyst/builds/124707853

Could you please run [this command](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md#codestyle)?

**Comment by i-zaitsev** [comment]
@TezRomacH ah ok, makes sense. I'll review the style and apply the command then.

**Comment by i-zaitsev** [comment]
BTW, got the following output after `make codestyle catalyst/data/datasets.py`:
```
./catalyst/contrib/models/sequential.py:22:5: C901 'SequentialNet.__init__' is too complex (11)
./catalyst/contrib/models/sequential.py:49:13: E731 do not assign a lambda expression, use a def
./catalyst/utils/config.py:194:1: C901 'parse_config_args' is too complex (12)
./catalyst/dl/meters/apmeter.py:34:5: C901 'APMeter.add' is too complex (12)
./catalyst/rl/scripts/run_samplers.py:143:1: C901 'main' is too complex (12)
4     C901 'SequentialNet.__init__' is too complex (11)
1     E731 do not assign a lambda expression, use a def
5
...
```
But I guess it is not related to this PR, and it somehow started recursively checking other files.

**Comment by TezRomacH** [comment]
Some troubles with the code-style happened, trying to fix it in #335

**Pull request merged**